src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -127,18 +135,20 @@ public boolean hasTeacher(Teacher teacher) {
     public void addTeacher(Teacher teacher) {
         addressBook.addTeacher(teacher);
         updateFilteredTeacherList(PREDICATE_SHOW_ALL_TEACHERS);
+        this.history.push(new AddressBook(this.addressBook));
     }
 
     @Override
     public void deleteTeacher(Teacher target) {
         addressBook.removeTeacher(target);
+        this.history.push(new AddressBook(this.addressBook));
     }
 
     @Override
     public void setTeacher(Teacher target, Teacher editedTeacher) {
         requireAllNonNull(target, editedTeacher);
-
         addressBook.setTeacher(target, editedTeacher);
+        this.history.push(new AddressBook(this.addressBook));
```

Comment:
I think it doesn't make sense to undo a filter/find command, since the user can just relist everything (and also it's quite hard to implement). That being said, let me test if undo works with the medical history command.